File: python/grass_sl_insert.py
import random
import pprint

class node:
	def __init__(self, nxt = None):
		self.next = nxt

##################
# accessory func #
##################
def create_list(size):
	root = None
	tl = root
	for i in range(1, size):
		new_node = node()
		if root is None:
			root = new_node
			tl = root
		else:
			tl.next = new_node
			tl = new_node
	return root

##################
#  end acc func  #
##################

def sl_insert(lst, elt):
	if lst is None:
		return elt
	else:
		nondet = 1
		curr = lst
		while nondet > 0 and curr.next is not None:
			nondet = random.randint(-2, 5)
			curr = curr.next
		curr_next = curr.next
		elt.next = curr_next
		curr.next = elt
		return lst

def main():
	root = create_list(10)
	elt = node()
	insert(root, elt)
	insert(None, elt)

if __name__=="__main__":
	main()

File: java/retrieve heap/src/testsuites/Grass_Sl_Insert.java
package testsuites;

import java.util.Random;

public class Grass_Sl_Insert {
    static GSLNode create_list(int size){
        GSLNode root = null;
        GSLNode tl = root;
        for(int i = 0; i < size; i++){
            GSLNode new_node = new GSLNode();
            new_node.next = null;
            if(root == null){
                root = new_node;
                tl = root;
            }else{
                tl.next = new_node;
                tl = new_node;
            }
        }

        return root;
    }

    static GSLNode sl_insert(GSLNode lst, GSLNode elt)
    {
        Random rand = new Random();
        //pre
        if (lst == null) {
            //post
            return elt;
        } else {
            int nondet = rand.nextInt(10)%10 - 5;
            GSLNode curr = lst;
            while(nondet> 0 && curr.next != null)
            {
                //loop
                nondet = rand.nextInt(10)%10 - 5;
                curr = curr.next;
            }
            GSLNode curr_next = curr.next;
            elt.next = curr_next;
            curr.next = elt;
            //post
            return lst;
        }
    }

    public static void main(String args[]){
        GSLNode lst = create_list(10);
        GSLNode elt = new GSLNode();
        elt.next = null;
        GSLNode res = sl_insert(lst, elt);
        sl_insert(null, elt);
    }

}

class GSLNode{
    GSLNode next;
}
